### PR TITLE
[WIP] Add a fetcher for APS

### LINKED ISSUE #818: Full text search for American Physical Society (APS) journals
Currently the full text search for articles published in [APS journals](http://journals.aps.org/) fails.

It would be great if it could be supported in a future release.

### PR DESCRIPTION
**Pull request by augustjanse**
Added a fetcher for APS, fixes #818, part of #2581. Based on the ScienceDirect fetcher.

I consider #818 fixed because this is all that can be done without an API key or token, and I hardly think any terms would let them be distributed with this software. I'm not exactly sure what your standards are on that, though.

- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not: Issue created at <https://github.com/JabRef/user-documentation/issues>.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - Filenames of external files can no longer contain curly braces. [#5926](https://github.com/JabRef/jabref/pull/5926)
 - We made the filters more easily accessible in the integrity check dialog. [#5955](https://github.com/JabRef/jabref/pull/5955)
 - We reimplemented and improved the dialog "Customize entry types" [#4719](https://github.com/JabRef/jabref/issues/4719)
+- We added an APS fetcher. [#818](https://github.com/JabRef/jabref/issues/818)
```
Comment (koppor) [inline CHANGELOG.md]: Can you also link APS please? I think, this is not a common term and it would help readers to get what it is. Maybe also write the long form of APS and add a link ^^.

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import org.jabref.model.entry.identifier.DOI;
+
+import kong.unirest.Unirest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * FulltextFetcher implementation that attempts to find a PDF URL at APS.
+ *
+ * @see 'http://harvest.aps.org/docs/harvest-api'
```
[the remark this answers is not in the document]
Comment (augustjanse) [reply src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: Fixed this like that.

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    // The actual API needs either an API key or a header. This is a workaround.
+    private static final String DOI_URL = "https://www.doi.org/";
+    private static final String PDF_URL = "https://journals.aps.org/prl/pdf/";
+
+    @Override
+    public Optional<URL> findFullText(BibEntry entry) throws IOException {
+        Objects.requireNonNull(entry);
+
+        Optional<DOI> doi = entry.getDOI();
```
[the remark this answers is not in the document]
Comment (augustjanse) [reply src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: Rewrote it like that. A bit unsure about throwing an unchecked exception, is that fine? I guess those are supposed to be handled elsewhere, since the method interface throws them.

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    public TrustLevel getTrustLevel() {
+        return TrustLevel.PUBLISHER;
+    }
+
+    /**
+     * Convert a DOI into an appropriate APS id.
+     *
+     * @param doi A case insensitive DOI
+     * @return A DOI cased as APS likes it
+     * @throws IOException
```
[the remark this answers is not in the document]
Comment (augustjanse) [reply src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: Fixed this.

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        // so we follow DOI.org redirects to get the proper id.
+        // https://stackoverflow.com/a/5270162/1729441
+        String doiRequest = DOI_URL + doi;
+        URLConnection con = new URL(doiRequest).openConnection();
+        con.connect();
+
+        // Throws FileNotFoundException if DOI doesn't exist
+        InputStream is = con.getInputStream();
+
+        return con.getURL().toString().split("abstract/")[1];
```
[the remark this answers is not in the document]
Comment (augustjanse) [reply src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: Rewrote it like that. I threw an exception that was already being caught instead of returning null, hope that's fine.

File: src/test/java/org/jabref/logic/importer/fetcher/ApsFetcherTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    @BeforeEach
+    void setUp() {
+        finder = new ApsFetcher();
+        entry = new BibEntry();
+    }
+
+    @Test
+    @DisabledOnCIServer("CI server is blocked")
+    void findOpenAccess() throws IOException {
+        entry.setField(StandardField.DOI, "10.1103/PhysRevLett.116.061102");
```
[the remark this answers is not in the document]
Comment (augustjanse) [reply src/test/java/org/jabref/logic/importer/fetcher/ApsFetcherTest.java]: Did that, not exactly sure about the style

File: src/test/java/org/jabref/logic/importer/fetcher/ApsFetcherTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    private BibEntry entry;
+
+    @BeforeEach
+    void setUp() {
+        finder = new ApsFetcher();
+        entry = new BibEntry();
+    }
+
+    @Test
+    @DisabledOnCIServer("CI server is blocked")
```
[the remark this answers is not in the document]
Comment (augustjanse) [reply src/test/java/org/jabref/logic/importer/fetcher/ApsFetcherTest.java]: My bad, sorry!

File: CHANGELOG.md (hunk 1 above)
Comment (augustjanse) [reply CHANGELOG.md]: Done

**Comment by augustjanse** [comment]
Thanks for your feedback! Please see the new commits.

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * FulltextFetcher implementation that attempts to find a PDF URL at APS. Also see the <a
+ * href="https://harvest.aps.org/docs/harvest-api">API</a>, although it isn't currently used.
+ */
+public class ApsFetcher implements FulltextFetcher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ApsFetcher.class);
+
+    // The actual API needs either an API key or a header. This is a workaround.
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: Not sure what "workaround" refers to. Maybe, the comment should be moved to the actual code - and not to constants.

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import org.jabref.logic.importer.FulltextFetcher;
+import org.jabref.model.entry.BibEntry;
+
+import kong.unirest.Unirest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * FulltextFetcher implementation that attempts to find a PDF URL at APS. Also see the <a
+ * href="https://harvest.aps.org/docs/harvest-api">API</a>, although it isn't currently used.
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: What do you mean by "used" at the end? Is the link for more information for another implementation alternative, which you did not choose. Could you give reasons why? Maybe, one sentence is enough. -- Then separate that from the first sentence.

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        return entry.getDOI().map(doi -> {
+            String id;
+            try {
+                id = getId(doi.getDOI());
+            } catch (FileNotFoundException e) {
+                // No such DOI
+                return null;
+            } catch (IOException e) {
+                // Handle IOExceptions elsewhere
+                throw new UncheckedIOException(e);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: This will crash complete JabRef - or did you catch it out of the stream thing?

Proposal: Log the exception and return `null`. -- Although `null` is not good - with the current code, it seems that `null` has to be used.

Did my proposal from yesterday not work out? ^^

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+                try {
+                    return new URL(pdfRequest);
+                } catch (MalformedURLException e) {
+                    LOGGER.info("APS returned malformed URL, cannot find PDF.");
+                    return null;
+                }
+            } else {
+                return null;
+            }
+        });
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: Please add `filter(result -> result != null)` to avoid returning an Optional containing a null value - or does this not happen? Did you add test cases for some of the `null` branches? ^^ - Maybe, you need to use [wiremock](http://wiremock.org/).

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    private String getId(String doi) throws IOException {
+        // DOI is not case sensitive, but the id for the PDF URL is,
+        // so we follow DOI.org redirects to get the proper id.
+        // https://stackoverflow.com/a/5270162/1729441
+        String doiRequest = DOI_URL + doi;
+        URLConnection con = new URL(doiRequest).openConnection();
+        con.connect();
+
+        // Throws FileNotFoundException if DOI doesn't exist
+        con.getInputStream();
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: Shouldn't the `getInputStream()` be wrapped in try-with-resources?

Shouldn't the `openConnection()` be wrapped in try-with-resources?

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+        // Expected URL example:
+        // https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.116.061102
+        // Expected parts: "https://journals.aps.org/prl/", "10.1103/PhysRevLett.116.061102"
+        String[] urlParts = con.getURL().toString().split("abstract/");
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: Thank you for the comment. Think, I would also have implemented it like that. Others would have used Regular Expressions. However, since this is simple, it is good to go.

ALTERNATIVE code - fosters reuse -- however, appears to be more complex

```java
return org.jabref.model.entry.identifier.DOI.findInText(con.getURL.toString()).map(DOI::getDOI).orElseThrow(ew FileNotFoundException("Unexpected URL received"));
```

ALTERNATIVE 2: Why not returning `Optional<DOI>` here and NOT throwing any exception? In the code above, you can then continue with `map` instead of throwing unchecked exceptions.

File: src/test/java/org/jabref/logic/importer/fetcher/ApsFetcherTest.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        BibEntry lowercase = new BibEntry();
+        lowercase.setField(StandardField.DOI, "10.1103/physrevlett.124.029002");
+        Optional<URL> pdfUrl2 = Optional.of(new URL("https://journals.aps.org/prl/pdf/10.1103/PhysRevLett.124.029002"));
+
+        // Article behind paywall returns empty
+        BibEntry unauthorized = new BibEntry();
+        unauthorized.setField(StandardField.DOI, "10.1103/PhysRevLett.89.127401");
+
+        // Unavailable article returns empty
+        BibEntry notFoundByDoi = new BibEntry();
```
Comment (koppor) [inline src/test/java/org/jabref/logic/importer/fetcher/ApsFetcherTest.java]: USe

```java
new BibEntry().withField(StandardField.DOI, "10.1016/j.aasri.2014.0559.002");
```

Then you can also directly put it into line 52 - and add the comment above line 52 -- save the variable

This can also be done for the other test cases... try to only return the Stream and put the expected results and requests there.

```java
return Stream.of(
    // Standard DOI works
    Arguments.of(
      Optional.of(new URL("https://journals.aps.org/prl/pdf/10.1103/PhysRevLett.116.061102")),
      new BibEntry().withField(StandardField.DOI, "10.1103/PhysRevLett.116.061102)),

    // DOI in lowercase works
    Arguments.of(
...
```

**Comment by koppor** [comment]
The codacy report is because of a [PMD](https://pmd.github.io/) false positive. See https://github.com/pmd/pmd/issues/1175 for details.

File: src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+    // The actual API needs either an API key or a header. This is a workaround.
+    private static final String DOI_URL = "https://www.doi.org/";
+    private static final String PDF_URL = "https://journals.aps.org/prl/pdf/";
+
+    @Override
+    public Optional<URL> findFullText(BibEntry entry) throws IOException {
+        Objects.requireNonNull(entry);
+
+        return entry.getDOI().map(doi -> {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java]: The code is more complicated due to functional style (which needs to catch exceptions etc). Thus either use a normal version (with quick exit ala `if(entry.getDoi().isEmpty()) return empty() else ... `) or alternatively try to reduce the exceptions thrown. For example, the FileNotFoundEx is only thrown to signal the the id is empty (if I'm not mistaken). A prefered option would be to make `getId` return `optional<string>` and than use `flatMap` instead of `map`.

In summary, the functional style is ok if the code could be refactored such that the following works:
````
return entry.getDOI().flatMap(this::determineId).flatMap(this::determinePdfUrl);

```

**Comment by koppor** [comment]
@tobiasdiez Reading https://github.com/augustjanse/jabref/blob/7a855dc0fc28df6ac546ae289367522c47dad643/src/main/java/org/jabref/logic/importer/fetcher/ApsFetcher.java#L26 and the description of the PR, we don't need any API key. I think, this is a good thing, isn't it? Is your idea to rewrite the code to rely on an API key?

**Comment by tobiasdiez** [comment]
Yes, the code works right now without an API key. However, it's kind of hacky and relies on a certain url format scheme which may break at any point.

I think it doesn't hurt to ask if we can get access to the official API. If it doesn't work, then we can still go ahead with the current workaround.

**Comment by augustjanse** [comment]
Thanks for your further feedback. This functional style is a bit new to me, so excuse me for falling into what I can only guess are common pitfalls, but your comments are very helpful. I'll try to do the changes in a few days.

Regarding the API: @tobiasdiez is right about the current solution being a hack. However, my understanding is that a `FulltextFetcher` must return a direct link to the PDF that doesn't need any headers with the request. If so, I don't think the API can supply such a link even with an API key or a CHORUS token. My comment might have been misleading on this part.

(The ScienceDirect fetcher got such a link by first making a request with the key in its header, and then extracting the direct link from the response. But from what I can see, it's not anywhere in any of the formats.)

**Comment by augustjanse** [comment]
Thanks for your further feedback. This functional style is a bit new to me, so excuse me for falling into what I can only guess are common pitfalls, but your comments are very helpful. I'll try to do the changes in a few days.

Regarding the API: @tobiasdiez is right about the current solution being a hack. However, my understanding is that a FulltextFetcher must return a direct link to the PDF that doesn't need any headers with the request. If so, I don't think the API can supply such a link even with an API key or a CHORUS token. My comment might have been misleading on this part.

(The ScienceDirect fetcher got such a link by first making a request with the key in its header, and then extracting the direct link from the response. But from what I can see, it's not anywhere in any of the formats.)

**Comment by sambo57u** [comment]
I am not sure this is relevant here but fetching bibliography entries from APS using DOI does not bring their page number. This has been a problem for a few years and I contacted APS and they said they are working on it and it was due to some indexing issue. That was a few years ago and nothing happened. e.g. try fetching 10.1103/PhysRevC.101.014611

**Comment by koppor** [comment]
Decision: Go without key; in a follow-up PR switch to new AP

Use `withField` instead of `setField`.
